File: src/main/java/com/mparticle/kits/CustomDataParser.java
package com.mparticle.kits;

import com.apptentive.android.sdk.ApptentiveLog;

import java.util.HashMap;
import java.util.Map;

final class CustomDataParser {
    public static Map<String, Object> parseCustomData(Map<String, String> map) {
        Map<String, Object> res = new HashMap<>();
        for (Map.Entry<String, String> e : map.entrySet()) {
            res.put(e.getKey(), parseValue(e.getValue()));
        }
        return res;
    }

    public static Object parseValue(String value) {
        try {
            return value != null ? parseValueGuarded(value) : null;
        } catch (Exception e) {
            ApptentiveLog.e(e, "Unable to parse value: '%s'", value);
            return value;
        }
    }

    private static Object parseValueGuarded(String value) {
        // check for boolean
        if ("true".equalsIgnoreCase(value) || "false".equalsIgnoreCase(value)) {
            return Boolean.parseBoolean(value);
        }

        // check for number
        Number number = StringUtils.tryParseNumber(value);
        if (number != null) {
            return number;
        }

        // default to the original value
        return value;
    }
}

File: src/main/kotlin/com/mparticle/kits/CustomDataParser.kt
package com.mparticle.kits

import android.util.Log
import kotlin.Any
import kotlin.Exception
import kotlin.String

internal object CustomDataParser {
    fun parseCustomData(map: Map<String, String?>): Map<String, Any?> {
        val res = HashMap<String, Any?>()
        for ((key, value) in map) {
            res[key] = parseValue(value)
        }
        return res
    }

    @JvmStatic
    fun parseValue(value: String?): Any? {
        return try {
            if (value != null) parseValueGuarded(value) else null
        } catch (e: Exception) {
            Log.e("MParticle-Util", "Unable to parse value: $value")
            value
        }
    }

    private fun parseValueGuarded(value: String): Any {
        // check for boolean
        if ("true".equals(value, true) || "false".equals(value,true)) {
            return value.toBoolean()
        }

        // check for number
        val number = StringUtils.tryParseNumber(value)
        return number ?: value

        // default to the original value
    }
}